File: AoC_2025/day_01.c
#include <stdio.h>
#include <stdlib.h>

int part1(int *data, int size){
    int zeros = 0;
    int position = 50;
    for (int i = 0; i < size; i++){
        int step = data[i];
        if (abs(data[i]) >= 100){
            step = abs(data[i]) % 100 * (data[i] < 0 ? -1 : 1);
        }
        position += step;
        if (position < 0) position += 100;
        if (position >= 100) position -= 100;
        if (position == 0) zeros++;
    }
    return zeros;
}

int part2(int *data, int size){
    int zeros = 0;
    int position = 50;
    for (int i = 0; i < size; i++){
        int step = data[i];
        int previous_position = position;
        if (abs(data[i]) >= 100){
            zeros += abs(data[i]) / 100;
            step = abs(data[i]) % 100 * (data[i] < 0 ? -1 : 1);
        }
        position += step;
        if (position < 0) {
            position += 100;
            if (position != 0 && previous_position != 0)
                zeros++;
        }
        if (position >= 100) {
            position -= 100;
            if (position != 0 && previous_position != 0)
                zeros++;
        }
        if (position == 0) zeros++;
    }
    return zeros;
}

int main(){
    FILE *f = fopen("day_01.txt", "r");
    if (!f) {
        perror("Failed to open file");
        return 1;
    }

    int capacity = 256;
    int size = 0;
    int *data = malloc(capacity * sizeof(int));
    if (!data) {
        perror("Memory allocation failed");
        fclose(f);
        return 1;
    }

    char line[64];

    while (fgets(line, sizeof(line), f)) {
        if (size >= capacity) { // Resize array if needed
            capacity *= 2;
            data = realloc(data, capacity * sizeof(int));
            if (!data) {
                perror("Memory reallocation failed");
                fclose(f);
                return 1;
            }
        }

        char direction = line[0];
        int value = atoi(&line[1]);
        data[size++] = (direction == 'L') ? -value : value;
    }

    fclose(f);
    printf("Part 1: %d\n", part1(data, size));
    printf("Part 2: %d\n", part2(data, size));
    free(data);
    return 0;
}

File: AoC_2025/day_01.py
import os

def part1(data):
    position = 50
    zeros = 0
    for move in data:
        if abs(move) >= 100:
            move = abs(move) % 100 * (1 if move >= 0 else -1)
        position += move
        if position < 0:
            position += 100
        if position > 99:
            position -= 100
        if position == 0:
            zeros += 1
    return zeros

def part2(data):
    position = 50
    zeros = 0
    for move in data:
        if abs(move) >= 100:
            zeros += abs(move) // 100
            move = abs(move) % 100 * (1 if move >= 0 else -1)
        prev = position
        position += move

        if position < 0:
            position += 100
            if position != 0 and prev != 0:
                zeros += 1

        if position > 99:
            position -= 100
            if position != 0 and prev != 0:
                zeros += 1

        if position == 0:
            zeros += 1 
    return zeros

if __name__ == "__main__":
    os.chdir(os.path.dirname(os.path.abspath(__file__)))
    with open("day_01.txt", "r") as f:
        data = [int(line[1:]) * (-1 if line[0] == "L" else 1) for line in f.read().splitlines()]
        print("Answer: ", part1(data))
        print("Answer: ", part2(data))
    